test/e2e/storage/testsuites/volumemode.go

```diff
@@ -0,0 +1,433 @@
+	. "github.com/onsi/gomega"
+
+	"k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+const (
+	noProvisioner = "kubernetes.io/no-provisioner"
+	pvNamePrefix  = "pv"
+)
+
+type volumeModeTestSuite struct {
+	name         string
+	testPatterns []testpatterns.TestPattern
+}
+
+func InitVolumeModeTestSuite() TestSuite {
+	return &volumeModeTestSuite{
+		name: "volumeMode [Feature:BlockVolume]",
+		testPatterns: []testpatterns.TestPattern{
+			testpatterns.DefaultFsImmediateBindPreprovisionedPV,
+			testpatterns.DefaultFsImmediateBindDynamicPV,
+			testpatterns.BlockImmediateBindPreprovisionedPV,
+			testpatterns.BlockImmediateBindDynamicPV,
+		},
+	}
+}
+
+func (t *volumeModeTestSuite) GetName() string {
+	return t.name
+}
+
+func (t *volumeModeTestSuite) GetTestPatterns() []testpatterns.TestPattern {
+	return t.testPatterns
+}
+
+func (t *volumeModeTestSuite) SkipUnsupportedTest(tp testpatterns.TestPattern, drv drivers.TestDriver) {
+}
+
+func (t *volumeModeTestSuite) InitTestResource() TestResource {
+	return &volumeModeTestResource{}
+}
+
+func (t *volumeModeTestSuite) CreateTestInput(tp testpatterns.TestPattern, tr TestResource) TestInput {
+	if r, ok := tr.(*volumeModeTestResource); ok {
+		drv := r.driver
+		di := drv.GetDriverInfo()
+		f := di.F
+
+		return volumeModeTestInput{
+			f:                f,
+			sc:               r.sc,
+			pvc:              r.pvc,
+			pv:               r.pv,
+			testVolType:      tp.VolType,
+			nodeName:         di.Config.ClientNodeName,
+			volMode:          tp.VolMode,
+			isBlockSupported: di.IsBlockSupported,
+		}
+	}
+
+	framework.Failf("Fail to convert testResource to volumeModeTestResource")
+	return nil
+}
+
+func (t *volumeModeTestSuite) GetTestFunc(tp testpatterns.TestPattern, drv drivers.TestDriver) func(*TestInput) {
+	di := drv.GetDriverInfo()
+	isBlockSupported := di.IsBlockSupported
+	volMode := tp.VolMode
+	volType := tp.VolType
+
+	switch volType {
+	case testpatterns.PreprovisionedPV:
+		if volMode == v1.PersistentVolumeBlock && !isBlockSupported {
+			return generateVolumeModeTestFunc(testVolumeModeFailForPreprovisionedPV)
+		} else {
+			return generateVolumeModeTestFunc(testVolumeModeSuccessForPreprovisionedPV)
+		}
+	case testpatterns.DynamicPV:
+		if volMode == v1.PersistentVolumeBlock && !isBlockSupported {
+			return generateVolumeModeTestFunc(testVolumeModeFailForDynamicPV)
+		} else {
+			return generateVolumeModeTestFunc(testVolumeModeSuccessForDynamicPV)
+		}
+	default:
+		framework.Failf("Volume mode test doesn't support volType: %v", volType)
+	}
+	return nil
+}
+
+func generateVolumeModeTestFunc(testFunc func(*volumeModeTestInput)) func(*TestInput) {
+	return func(ti *TestInput) {
```

Comment:
Seems like all test suites have to copy a similar method. Is there anyway we can consolidate this?

Reply:
@msau42 

Thank you for your suggestion. I also would like to make it a common method, and tried something like below to replace ```generateVolumeModeTestFunc``` to more general one,  but it doesn't work. Because they seem to access invalid memory in parse phase. Any idea or advice for this?

```
func convertToReflectValue(ti *TestInput) reflect.Value {
    var v reflect.Value

    switch t := (*ti).(type) {
    case volumesTestInput:
        v = reflect.ValueOf(&t)
    case volumeIOTestInput:
        v = reflect.ValueOf(&t)
    case subPathTestInput:
        v = reflect.ValueOf(&t)
    case volumeModeTestInput:
        v = reflect.ValueOf(&t)
    default:
        framework.Failf("Fail to convert testInput %v", ti)
    }

    return v
}

func generateTestFunc(testFunc interface{}) func(*TestInput) {
    return func(ti *TestInput) {
        Context((*ti).GetTestMetadataString(), func() {
            var arg reflect.Value

            BeforeEach(func() {
                arg = convertToReflectValue(ti)
            })

            doTest := func(a *reflect.Value) {
                fVal := reflect.ValueOf(testFunc)
                fType := fVal.Type()
                // Check if testFunc is a function that takes exact one argument and return no value
                // ex) func fooTest(*fooTestInput)
                if fType.Kind() != reflect.Func || fType.NumIn() != 1 || fType.NumOut() != 0 {
                    framework.Failf("testFunc %v isn't a suitable function to test", testFunc)
                }

                fVal.Call([]reflect.Value{*a})
            }

            // Call test function
            doTest(&arg)
        })
    }
}
```

Reply:
Hm, does the framework need to pass the input back and forth with the test?  Could the test function take the test pattern as an input, and have a beforeeach() that generates in the input?

Reply:
Thank you for your advice. 

It seems that I managed to make it work with a bit another approach.
It won't be optimal, but it should improve the readability of the code. 
I will work on other minor fixes and will commit the code for review.

Anyway, thank you so much for your suggestions and advices! The codes become much better.